File: src/lhp/api/schemas/ai.py
"""Schemas for AI assistant endpoints."""

from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel


class SessionMode(str, Enum):
    """Session permission mode."""

    agent = "agent"
    chat = "chat"


class AIStatusResponse(BaseModel):
    """Response for GET /api/ai/status."""

    available: bool
    url: Optional[str] = None
    auth_required: bool = False
    config: Optional[dict[str, Any]] = None


class AISessionCreateRequest(BaseModel):
    """Request body for POST /api/ai/session."""

    mode: SessionMode = SessionMode.agent


class AISessionResponse(BaseModel):
    """Response for POST /api/ai/session."""

    session_id: str
    project_root: str
    mode: str = "agent"


class AISessionDeleteResponse(BaseModel):
    """Response for DELETE /api/ai/session/{session_id}."""

    deleted: bool


class AISessionSummary(BaseModel):
    """Summary of one OpenCode session."""

    id: str
    title: Optional[str] = None
    created_at: Optional[str] = None


class AISessionListResponse(BaseModel):
    """Response for GET /api/ai/sessions."""

    sessions: list[AISessionSummary]


class AIMessagePart(BaseModel):
    """A single part in a message sent to OpenCode."""

    type: str = "text"
    text: str = ""


class AIMessageRequest(BaseModel):
    """Request body for POST /api/ai/session/{id}/message."""

    parts: list[AIMessagePart]


class AIGenerateTitleRequest(BaseModel):
    """Request body for POST /api/ai/session/{id}/generate-title."""

    parts: list[AIMessagePart]


class AIConfigSummary(BaseModel):
    """Response for GET /api/ai/config — allowed providers/models."""

    provider: str
    model: str
    allowed_models: dict[str, list[str]]


class AIConfigUpdateRequest(BaseModel):
    """Request body for POST /api/ai/config."""

    provider: str
    model: str

File: web_app/src/api/ai.ts
/**
 * API client for AI assistant endpoints.
 *
 * All AI traffic routes through /api/ai/* — the frontend never
 * connects to OpenCode directly.
 */

import { fetchApi } from './client'
import type { AIStatus, SessionMode } from '../types/chat'

// ── Status & config ──────────────────────────────────────────

export interface AIConfigSummary {
  provider: string
  model: string
  allowed_models: Record<string, string[]>
}

export function fetchAIStatus(): Promise<AIStatus> {
  return fetchApi<AIStatus>('/ai/status')
}

export function fetchAIConfig(): Promise<AIConfigSummary> {
  return fetchApi<AIConfigSummary>('/ai/config')
}

export function updateAIConfig(provider: string, model: string): Promise<AIConfigSummary> {
  return fetchApi<AIConfigSummary>('/ai/config', {
    method: 'POST',
    body: JSON.stringify({ provider, model }),
  })
}

// ── Session management ───────────────────────────────────────

export interface AISessionResponse {
  session_id: string
  project_root: string
  mode: string
}

export function createAISession(mode: SessionMode = 'agent'): Promise<AISessionResponse> {
  return fetchApi<AISessionResponse>('/ai/session', {
    method: 'POST',
    body: JSON.stringify({ mode }),
  })
}

export function listAISessions(): Promise<unknown[]> {
  return fetchApi<unknown[]>('/ai/sessions')
}

export function deleteAISession(sessionId: string): Promise<{ deleted: boolean }> {
  return fetchApi(`/ai/session/${sessionId}`, { method: 'DELETE' })
}

// ── Messages ─────────────────────────────────────────────────

export interface AIMessagePart {
  type: string
  text: string
}

export function sendAIMessage(
  sessionId: string,
  parts: AIMessagePart[],
): Promise<{ ok: boolean }> {
  return fetchApi<{ ok: boolean }>(`/ai/session/${sessionId}/message`, {
    method: 'POST',
    body: JSON.stringify({ parts }),
  })
}

export function getSessionMessages(sessionId: string): Promise<unknown[]> {
  return fetchApi<unknown[]>(`/ai/session/${sessionId}/messages`)
}

// ── Title generation ─────────────────────────────────────────

/** Fire-and-forget: ask the backend to generate an LLM title for the session. */
export function generateSessionTitle(
  sessionId: string,
  parts: AIMessagePart[],
): Promise<{ ok: boolean; title?: string }> {
  return fetchApi<{ ok: boolean; title?: string }>(
    `/ai/session/${sessionId}/generate-title`,
    {
      method: 'POST',
      body: JSON.stringify({ parts }),
    },
  )
}

// ── Cancel ───────────────────────────────────────────────────

export function cancelAIGeneration(sessionId: string): Promise<{ ok: boolean }> {
  return fetchApi<{ ok: boolean }>(`/ai/session/${sessionId}/cancel`, {
    method: 'POST',
  })
}

// ── Question tool response ──────────────────────────────────

/** Fetch pending question requests (for catching up after SSE reconnect). */
export function fetchPendingQuestions(): Promise<Array<{ id: string; sessionID: string }>> {
  return fetchApi<Array<{ id: string; sessionID: string }>>('/ai/questions')
}

/**
 * Reply to a question tool call.  OpenCode's question tool blocks
 * until it receives a structured answer via this endpoint.
 *
 * @param answers — parallel array matching `questions[]`.  Each
 *   element is a `string[]` of selected option labels (even for
 *   single-select, wrap in an array: `["Option A"]`).
 */
export function replyToQuestion(
  sessionId: string,
  requestId: string,
  answers: string[][],
): Promise<{ ok: boolean }> {
  return fetchApi<{ ok: boolean }>(
    `/ai/session/${sessionId}/question/${requestId}/reply`,
    { method: 'POST', body: JSON.stringify({ answers }) },
  )
}

/** Dismiss a question without answering. */
export function rejectQuestion(
  sessionId: string,
  requestId: string,
): Promise<{ ok: boolean }> {
  return fetchApi<{ ok: boolean }>(
    `/ai/session/${sessionId}/question/${requestId}/reject`,
    { method: 'POST' },
  )
}

// ── SSE event stream ─────────────────────────────────────────

/**
 * Connect to the SSE event stream for a session.
 *
 * Returns an EventSource that receives OpenCode events
 * proxied through the backend.  The caller is responsible
 * for closing it when done.
 */
export function connectAIEventStream(sessionId: string): EventSource {
  const baseUrl = import.meta.env.DEV ? '' : ''
  return new EventSource(`${baseUrl}/api/ai/session/${sessionId}/events`)
}

/**
 * Connect to the user-scoped SSE event stream (all sessions).
 *
 * Unlike connectAIEventStream, this receives events for ALL
 * sessions belonging to the user.  The frontend dispatches
 * events to the correct session tab using the sessionID field.
 */
export function connectAIUserEventStream(): EventSource {
  const baseUrl = import.meta.env.DEV ? '' : ''
  return new EventSource(`${baseUrl}/api/ai/events`)
}
